controller/api/destination/watcher/cluster_watcher.go

```diff
@@ -0,0 +1,320 @@
+					return
+				}
+				// If the zombie object is a `Secret` we can delete it.
+				secret, ok = tombstone.Obj.(*v1.Secret)
+				if !ok {
+					cs.log.Debugf("DeletedFinalStateUnknown contained object that is not a Secret %#v", obj)
+					return
+				}
+			}
+
+			clusterName, found := secret.GetLabels()[clusterNameLabel]
+			if !found {
+				cs.log.Tracef("Skipping Delete event for 'Secret' object %s/%s: missing \"%s\" label", secret.Namespace, secret.Name, clusterNameLabel)
+				return
+			}
+
+			cs.removeWatcher(clusterName)
+
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return cs, nil
+}
+
+// Get safely retrieves a watcher from the cache given a cluster name. It
+// returns the watcher, cluster configuration, if the entry exists in the cache.
+func (cs *ClusterStore) Get(clusterName string) (*EndpointsWatcher, string, string, bool) {
+	cs.RLock()
+	defer cs.RUnlock()
+	s, found := cs.store[clusterName]
+	return s.watcher, s.trustDomain, s.clusterDomain, found
+}
+
+// GetWatcher is a convenience method that given a cluster name only returns the
+// associated EndpointsWatcher if it exists in the cache.
+func (cs *ClusterStore) GetWatcher(clusterName string) (*EndpointsWatcher, bool) {
+	cs.RLock()
+	defer cs.RUnlock()
+	s, found := cs.store[clusterName]
+	return s.watcher, found
+}
+
+// GetLocalWatcher is a convenience method that retrieves the watcher associated
+// with the local cluster. Its existence is assumed.
+func (cs *ClusterStore) GetLocalWatcher() *EndpointsWatcher {
+	cs.RLock()
+	defer cs.RUnlock()
+	return cs.store[LocalClusterKey].watcher
+}
+
+// GetClusterConfig is a convenience method that given a cluster name retrieves
+// its associated configuration strings if the entry exists in the cache.
+func (cs *ClusterStore) GetClusterConfig(clusterName string) (string, string, bool) {
+	cs.RLock()
+	defer cs.RUnlock()
+	s, found := cs.store[clusterName]
+	return s.trustDomain, s.clusterDomain, found
+}
+
+// AddLocalWatcher adds a watcher to the cache using the local cluster key.
+func (cs *ClusterStore) AddLocalWatcher(stopCh chan<- struct{}, watcher *EndpointsWatcher, trustDomain, clusterDomain string) {
+	cs.Lock()
+	defer cs.Unlock()
+	cs.store[LocalClusterKey] = cluster{
+		watcher,
+		trustDomain,
+		clusterDomain,
+		stopCh,
+	}
+}
+
+// Len returns the number of entries in the cache
+func (cs *ClusterStore) Len() int {
+	cs.RLock()
+	defer cs.RUnlock()
+	return len(cs.store)
+}
+
+// removeWatcher is triggered by the cache's Secret informer when a secret is
+// removed. Given a cluster name, it removes the entry from the cache after
+// stopping the associated watcher.
+func (cs *ClusterStore) removeWatcher(clusterName string) {
+	cs.Lock()
+	defer cs.Unlock()
+	s, found := cs.store[clusterName]
+	if !found {
+		return
+	}
+	// For good measure, close the channel after stopping to ensure
+	// informers are shut down.
+	defer close(s.stopCh)
+	s.watcher.Stop(s.stopCh)
```

Comment:
In fact... I don't think writing to this channel does anything at all.  As far as I can tell, the informers are stopped when this channel is closed.  Which means we just have to close it, not write to it.

Reply:
Writing to it is supposed to signal shutdown, at least according to the [controller's Run() method](https://github.com/kubernetes/client-go/blob/c04e419b6ad1e94a7fd2d5cc9d8a2d5320470840/tools/cache/controller.go#L127-L130) (which [informers use to cache events](https://github.com/kubernetes/client-go/blob/c04e419b6ad1e94a7fd2d5cc9d8a2d5320470840/tools/cache/shared_informer.go#L503)).

In practice, the tests were only passing if we closed the channel... Both should be equivalent so I agree it's redundant to write & close. We can simply `close()` as you've suggested 👍🏻